File: JavaScript/Playlist.js

import Song from "./Song.js";

class Playlist {
    constructor(name, library) {
        this.name = name;
        this.songs = [];
        this.library = library;
    }

    addSongToPlaylist(title, artist, genre) {
        let song = this.library.songs.find(song => song.title === title && song.artist === artist);

        if (!song) {
            song = new Song(title, artist, genre);
            this.library.addSongToLibrary(song);
            console.log(`New song "${song.title}" added to playlist and library.`);
        }

        const existingSong = this.songs.find(s => s.id === song.id);

        if (!existingSong) {
            this.songs.push(song);
            console.log(`Song "${song.title}" added to playlist.`);
        } else {
            console.log(`Song "${song.title}" already exists in this playlist.`);
        }
    }

    searchSongsInPlaylist(searchTerm, attribute = 'title') {
        const results = this.songs.filter(song => {
            return song[attribute].toLowerCase().includes(searchTerm.toLowerCase());
        });

        if (results.length === 0) {
            console.log('No songs found for your search criteria.');
            return [];
        }
        return results;
    }

    updateSongInPlaylist(songId, title=null, artist=null, genre=null) {
        const song = this.songs.find(s => s.id === songId);

        if (song) {
            song.title = title || song.title;
            song.artist = artist || song.artist;
            song.genre = genre || song.genre;
            console.log(`Song with ID ${songId} updated successfully.`);
        } else {
            console.log(`Song with ID ${songId} not found in playlist.`);
        }
        return song;
    }

    removeSongFromPlaylist(songId) {
        const index = this.songs.findIndex(s => s.id === songId);

        if (index !== -1) {
            this.songs.splice(index, 1);
            console.log(`Song with ID ${songId} removed from playlist.`);
        } else {
            console.log(`Song with ID ${songId} not found in playlist.`);
        }
    }

    sortSongsInPlaylist(attributes = ['title']) {
        this.songs.sort((a, b) => {
            for (let attribute of attributes) {
                if (a[attribute] < b[attribute]) return -1;
                if (a[attribute] > b[attribute]) return 1;
            }
            return 0;
        });
    }

    displayPlaylist() {
        const playlistData = {
            name: this.name,
            songs: this.songs
        };

        console.log(`Playlist: ${this.name}`);
        if (this.songs.length === 0) {
            console.log('\tPlaylist currently has no songs.');
        } else {
            this.songs.forEach(song => {
                song.showSongDetails();
            });
        }
        return playlistData;
    }
}

export default Playlist;

File: Python/Playlist.py
from Song import Song

class Playlist:
    def __init__(self, name, library):
        self.name = name
        self.songs = []
        self.library = library

    def add_song_to_playlist(self, title, artist, genre):
        song = next((s for s in self.library.songs if s.title == title and s.artist == artist), None)

        if not song:
            song = Song(title, artist, genre)
            self.library.add_song_to_library(song)
            print(f'New song "{song.title}" added to playlist and library.')

        if not any(s.id == song.id for s in self.songs):
            self.songs.append(song)
            print(f'Song "{song.title}" added to playlist.')
        else:
            print(f'Song "{song.title}" already exists in this playlist.')

    def search_songs_in_playlist(self, search_term, attribute='title'):
        results = []
        if attribute == 'id':
            for song in self.songs:
                if int(search_term) == song.id:
                    results.append(song.show_song_details())
        else:
            results = [song.show_song_details() for song in self.songs if search_term.lower() in getattr(song, attribute).lower()]

        if not results:
            print('No songs found for your search criteria.')
            return []
        
        return results

    def update_song_in_playlist(self, song_id, title=None, artist=None, genre=None):
        song = next((s for s in self.songs if s.id == song_id), None)

        if song:
            song.title = title or song.title
            song.artist = artist or song.artist
            song.genre = genre or song.genre
            print(f'Song with ID {song_id} updated successfully.')
        else:
            print(f'Song with ID {song_id} not found in playlist.')

        return song

    def remove_song_from_playlist(self, song_id):
        song = next((s for s in self.songs if s.id == song_id), None)

        if song:
            self.songs.remove(song)
            print(f'Song with ID {song_id} removed from playlist.')
        else:
            print(f'Song with ID {song_id} not found in playlist.')

    def sort_songs_in_playlist(self, attributes=['title']):
        self.songs.sort(key=lambda song: [getattr(song, attr) for attr in attributes])

    def display_playlist(self):
        songs = []
        print(f'Playlist: {self.name}')
        if not self.songs:
            print('\tPlaylist currently has no songs.')
            songs.append('Playlist currently has no songs.')
        else:
            for song in self.songs:
                song_data = song.show_song_details()
                songs.append(song_data)

        return {'name': self.name, 'songs': songs}

